File: cli/src/org/scalablytyped/converter/cli/Tracing.scala
package org.scalablytyped.converter.cli

import org.scalablytyped.converter.Flavour
import org.scalablytyped.converter.Selection
import org.scalablytyped.converter.internal.IArray
import org.scalablytyped.converter.internal.InFolder
import org.scalablytyped.converter.internal.Json
import org.scalablytyped.converter.internal.constants
import org.scalablytyped.converter.internal.files
import org.scalablytyped.converter.internal.importer.Bootstrap
import org.scalablytyped.converter.internal.importer.ConversionOptions
import org.scalablytyped.converter.internal.importer.EnabledTypeMappingExpansion
import org.scalablytyped.converter.internal.importer.LibScalaJs
import org.scalablytyped.converter.internal.importer.LibTsSource
import org.scalablytyped.converter.internal.importer.PersistingParser
import org.scalablytyped.converter.internal.importer.Phase1ReadTypescript
import org.scalablytyped.converter.internal.importer.Phase2ToScalaJs
import org.scalablytyped.converter.internal.importer.PhaseFlavour
import org.scalablytyped.converter.internal.logging.*
import org.scalablytyped.converter.internal.maps.*
import org.scalablytyped.converter.internal.phases.PhaseListener
import org.scalablytyped.converter.internal.phases.PhaseRes
import org.scalablytyped.converter.internal.phases.PhaseRunner
import org.scalablytyped.converter.internal.phases.RecPhase
import org.scalablytyped.converter.internal.scalajs.Minimization
import org.scalablytyped.converter.internal.scalajs.Name
import org.scalablytyped.converter.internal.scalajs.PackageTree
import org.scalablytyped.converter.internal.scalajs.ParentsResolver
import org.scalablytyped.converter.internal.scalajs.Printer
import org.scalablytyped.converter.internal.scalajs.QualifiedName
import org.scalablytyped.converter.internal.scalajs.TreeScope
import org.scalablytyped.converter.internal.scalajs.Versions
import org.scalablytyped.converter.internal.ts.CalculateLibraryVersion.PackageJsonOnly
import org.scalablytyped.converter.internal.ts.PackageJson
import org.scalablytyped.converter.internal.ts.TsIdentLibrary

import scala.collection.immutable.SortedSet

object Tracing {
  private val inDirectory = os.pwd
  private val sourceOutputDir     = os.pwd / "generated-sources"
  private val paths          = new Paths(inDirectory)
  private val parseCachePath      = Some(files.existing(constants.defaultCacheFolder / "parse").toNIO)

  val logger: Logger[(Array[Logger.Stored], Unit)] =
    storing().zipWith(stdout.filter(LogLevel.warn))

  // Execution logger for comprehensive logging
  private val executionLogger = ExecutionLogger(inDirectory, sourceOutputDir)

  private val DefaultOptions = ConversionOptions(
    useScalaJsDomTypes = true,
    outputPackage = Name.typings,
    flavour = Flavour.Normal,
    enableScalaJsDefined = Selection.All,
    ignored = SortedSet("typescript"),
    stdLibs = SortedSet("es6"),
    versions = Versions(Versions.Scala3, Versions.ScalaJs1),
    expandTypeMappings = EnabledTypeMappingExpansion.DefaultSelection,
    enableLongApplyMethod = false,
    privateWithin = None,
    useDeprecatedModuleNames = false
  )

  def main(args: Array[String]): Unit = System.exit(mainNoExit())

  private def mainNoExit(): Int = {
    // Note: args parameter is currently unused but kept for future CLI argument parsing
    // Initialize execution logging
    executionLogger.initializeExecutionLog()

    try {
      executionLogger.logStep("Initializing converter and reading configuration")
      val packageJsonPath = paths.packageJson.getOrElse(sys.error(s"${inDirectory} does not contain package.json"))
      val nodeModulesPath = paths.node_modules.getOrElse(sys.error(s"${inDirectory} does not contain node_modules"))
      val packageJson: PackageJson = Json.force[PackageJson](packageJsonPath)
      executionLogger.logProgress(s"Found package.json at: $packageJsonPath")
      executionLogger.logProgress(s"Found node_modules at: $nodeModulesPath")

      val wantedLibs: SortedSet[TsIdentLibrary] = {
        val fromPackageJson = packageJson.allLibs(false, peer = true).keySet
        executionLogger.logProgress(s"Libraries found in package.json: ${fromPackageJson.map(_.value).mkString(", ")}")
        require(fromPackageJson.nonEmpty, "No libraries found in package.json")
        val ret = fromPackageJson -- DefaultOptions.ignoredLibs
        executionLogger.logProgress(s"Libraries after filtering ignored: ${ret.map(_.value).mkString(", ")}")
        require(ret.nonEmpty, "All libraries in package.json ignored")
        ret
      }

      executionLogger.logStep("Bootstrapping TypeScript environment")
      executionLogger.logProgress(s"Standard library configuration: ${DefaultOptions.stdLibs.mkString(", ")}")
      val bootstrapped: Bootstrap.Bootstrapped =
        Bootstrap.fromNodeModules(InFolder(nodeModulesPath), DefaultOptions, wantedLibs)
      executionLogger.logProgress(
        s"Bootstrap completed, found ${bootstrapped.libraryResolver.stdLib.libName.value} as stdlib"
      )

      val sources: Vector[LibTsSource] = {
        bootstrapped.initialLibs match {
          case Left(unresolved) =>
            executionLogger.logError(s"Failed to resolve initial libraries: ${unresolved.msg}")
            sys.error(unresolved.msg)
          case Right(initial) => initial
        }
      }

      executionLogger.logProgress(s"Initial sources from bootstrap: ${sources.map(_.libName.value).mkString(", ")}")
      executionLogger.logProgress(s"Converting ${sources.map(_.libName.value).mkString(", ")} to scalajs...")

      val cachedParser = PersistingParser(parseCachePath, bootstrapped.inputFolders, logger.void)

      executionLogger.logStep("Setting up TypeScript parsing phase")
      val phase1 = new Phase1ReadTypescript(
        resolve = bootstrapped.libraryResolver,
        calculateLibraryVersion = PackageJsonOnly,
        ignored = DefaultOptions.ignoredLibs,
        ignoredModulePrefixes = DefaultOptions.ignoredModulePrefixes,
        pedantic = false,
        parser = cachedParser,
        expandTypeMappings = DefaultOptions.expandTypeMappings
      )
      executionLogger.logProgress("TypeScript parsing phase configured")

      executionLogger.logStep("Setting up Scala.js conversion phase")
      val phase2 = new Phase2ToScalaJs(
        pedantic = false,
        scalaVersion = DefaultOptions.versions.scala,
        enableScalaJsDefined = DefaultOptions.enableScalaJsDefined,
        outputPkg = DefaultOptions.outputPackage,
        flavour = DefaultOptions.flavourImpl,
        useDeprecatedModuleNames = DefaultOptions.useDeprecatedModuleNames
      )
      executionLogger.logProgress("Scala.js conversion phase configured")

      executionLogger.logStep("Setting up flavour transformation phase")
      val phase3 = new PhaseFlavour(DefaultOptions.flavourImpl, maybePrivateWithin = DefaultOptions.privateWithin)
      executionLogger.logProgress(s"Flavour transformation phase configured: ${phase3}")

      executionLogger.logStep("Creating conversion pipeline")
      val pipeline: RecPhase[LibTsSource, LibScalaJs] = RecPhase[LibTsSource]
        .next(phase1, "typescript")
        .next(phase2, "scala.js")
        .next(phase3, DefaultOptions.flavour.toString)
      executionLogger.logProgress("Conversion pipeline created successfully")

      val NoListener: PhaseListener[LibTsSource] = (_, _, _) => ()

      executionLogger.logStep("Processing libraries through pipeline")
      executionLogger.logProgress(
        s"Processing ${sources.size} libraries: ${sources.map(_.libName.value).mkString(", ")}"
      )

      val importedLibs: Map[LibTsSource, PhaseRes[LibTsSource, LibScalaJs]] =
        sources
          .map(s => {
            executionLogger.logProgress(s"Processing library ${s.libName.value}")
            val result = PhaseRunner(pipeline, (_: LibTsSource) => logger.void, NoListener)(s)
            executionLogger.logProgress(s"Result for ${s.libName.value}: ${result.getClass.getSimpleName}")
            (s: LibTsSource) -> result
          })
          .toMap

      executionLogger.logProgress(s"Pipeline processing completed. Processed ${importedLibs.size} libraries")
      importedLibs.foreach { case (source, result) =>
        executionLogger.logProgress(s"  ${source.libName.value} -> ${result.getClass.getSimpleName}")
      }

      executionLogger.logStep("Processing results and generating Scala files")
      PhaseRes.sequenceMap(importedLibs.toSorted) match {
        case PhaseRes.Ok(LibScalaJs.Unpack(libs)) =>
          executionLogger.logProgress(s"Successfully processed ${libs.size} libraries")

          executionLogger.logStep("Creating global scope for all libraries")
          val globalScope = new TreeScope.Root(
            DefaultOptions.outputPackage,
            Name.dummy,
            libs.map { case (_, l) => (l.scalaName, l.packageTree) },
            logger.void,
            false
          )
          executionLogger.logProgress("Global scope created successfully")

          // Static minimization policy applied to every run
          val minimizes =
            List("quill", "clsx", "scroll-into-view-if-needed", "tabulator-tables").map(TsIdentLibrary.apply)
          val staticMinimize: Selection[TsIdentLibrary] = Selection.AllExcept(minimizes*)
          val staticMinimizeKeep: IArray[QualifiedName] = IArray.Empty

          // Compute shared keep index for minimization (only if needed)
          lazy val referencesToKeep: Minimization.KeepIndex = {
            val packagesWithShouldMinimize: IArray[(PackageTree, Boolean)] =
              IArray.fromTraversable(libs).map { case (s, l) => (l.packageTree, staticMinimize(s.libName)) }

            Minimization.findReferences(globalScope, staticMinimizeKeep, packagesWithShouldMinimize)
          }

          executionLogger.logStep("Generating source files")
          val allGeneratedFiles: Iterator[(os.Path, String)] = libs.iterator.flatMap { case (source, lib) =>
            val willMinimize = staticMinimize(source.libName)
            val treeToPrint =
              if (willMinimize) Minimization(globalScope, referencesToKeep, logger.void, lib.packageTree)
              else lib.packageTree

            val scalaFiles = Printer(
              globalScope,
              new ParentsResolver,
              treeToPrint,
              DefaultOptions.outputPackage,
              DefaultOptions.versions.scala
            )

            val targetFolder = sourceOutputDir / source.libName.value
            val minimizedMsg = if (willMinimize) "minimized " else ""
            executionLogger.logProgress(
              s"Writing ${minimizedMsg}${source.libName.value} (${scalaFiles.length} files) to $targetFolder..."
            )
            scalaFiles.map { case (relPath, content) =>
              (targetFolder / relPath, content)
            }.iterator
          }

          executionLogger.logStep("Writing files to disk")
          val writtenFiles = allGeneratedFiles.map { case (path, content) =>
            files.softWrite(path) { writer =>
              writer.write(content)
            }
            path
          }.toSet

          executionLogger.logProgress(
            s"Successfully generated ${writtenFiles.size} Scala source files to $sourceOutputDir"
          )
          executionLogger.finalizeExecutionLog(success = true)
          0

        case PhaseRes.Failure(errors) =>
          executionLogger.logError("Pipeline failed with errors:")
          errors.foreach {
            case (source, Left(throwable)) =>
              executionLogger.logError(s"  ${source.libName.value}: ${throwable.getMessage}", Some(throwable))
            case (source, Right(message)) =>
              executionLogger.logError(s"  ${source.libName.value}: $message")
          }
          executionLogger.finalizeExecutionLog(success = false)
          1

        case PhaseRes.Ignore() =>
          executionLogger.logWarning("Pipeline ignored all sources")
          executionLogger.finalizeExecutionLog(success = true)
          0
      }
    } catch {
      case ex: Exception =>
        executionLogger.logError(s"Execution failed with exception: ${ex.getMessage}", Some(ex))
        executionLogger.finalizeExecutionLog(success = false)
        throw ex
    }
  }
}

File: cli-ts/src/commands/tracing.ts
import * as path from "node:path";
import * as E from "fp-ts/Either";
import { pipe } from "fp-ts/function";
import * as fs from "fs-extra";
import { Flavour } from "@/Flavour.ts";
import { InFolder } from "@/internal/files.ts";
import {
	Bootstrap,
	type Bootstrapped,
	type Unresolved,
} from "@/internal/importer/Bootstrap.ts";
import { MockCalculateLibraryVersion } from "@/internal/importer/CalculateLibraryVersion.js";
import {
	ConversionOptions,
	Versions,
} from "@/internal/importer/ConversionOptions.ts";
import { NormalFlavourImpl } from "@/internal/importer/FlavourImpl.js";
import { Json, type LibTsSource } from "@/internal/importer/LibTsSource.js";
import { Phase1ReadTypescript } from "@/internal/importer/Phase1ReadTypescript.js";
import { Phase2ToScalaJs } from "@/internal/importer/Phase2ToScalaJs.js";
import { PhaseFlavour } from "@/internal/importer/PhaseFlavour.js";
import type { Logger } from "@/internal/logging/index.js";
import { CollectingPhaseListener } from "@/internal/phases/PhaseListener.js";
import {
	Formatters,
	Orderings,
	PhaseRunner,
} from "@/internal/phases/PhaseRunner.js";
import { RecPhase } from "@/internal/phases/RecPhase.js";
import { Selection } from "@/internal/Selection.ts";
import { Name } from "@/internal/scalajs/Name.ts";
import { PackageJson } from "@/internal/ts/PackageJson.js";
import { type TsIdentLibrary, TsParsedFile } from "@/internal/ts/trees.js";
import { ExecutionLogger } from "@/utils/ExecutionLogger.js";
import { Paths } from "@/utils/paths.js";
import { BaseCommand, type CommandOptions } from "./base-command.js";

/**
 * Main conversion command - equivalent to Scala Tracing.scala
 * Converts TypeScript definitions to Scala.js sources
 */
export class TracingCommand extends BaseCommand {
	private readonly inDirectory: string;
	private readonly sourceOutputDir: string;
	private readonly paths: Paths;
	private readonly executionLogger: ExecutionLogger;

	private readonly DefaultOptions = new ConversionOptions(
		true,
		Flavour.Normal,
		Name.typings,
		new Set(["es6"]),
		Selection.All(),
		Selection.All(),
		new Set(),
		new Versions(Versions.Scala3, Versions.ScalaJs1),
		false,
		undefined,
		false,
	);

	constructor(options: CommandOptions) {
		super(options);
		this.inDirectory = process.cwd();
		this.sourceOutputDir = path.resolve(
			options.output || "./generated-sources",
		);
		this.paths = new Paths(this.inDirectory);
		this.executionLogger = ExecutionLogger.create(
			this.inDirectory,
			this.sourceOutputDir,
		);
	}

	async execute(): Promise<void> {
		this.info("Starting TypeScript to Scala.js conversion...");

		try {
			// Initialize execution logging
			await this.executionLogger.initializeExecutionLog();

			// Step 0: Validate environment
			await this.validateEnvironment();

			const packageJsonPath = this.paths.packageJson;
			// TypeScript equivalent of: val packageJson: PackageJson = Json.force[PackageJson](packageJsonPath)
			const packageJson: PackageJson = Json.force(packageJsonPath, (obj) =>
				PackageJson.fromObject(obj),
			);

			// TypeScript equivalent of: val wantedLibs: SortedSet[TsIdentLibrary] = packageJson.allLibs(false, peer = true).keySet
			const wantedLibs: Set<TsIdentLibrary> = new Set(
				packageJson.allLibs(false, true).keys(),
			);

			this.executionLogger.logStep("Bootstrapping TypeScript environment");
			this.executionLogger.logProgress(
				`Standard library configuration: ${Array.from(this.DefaultOptions.stdLibs).join(", ")}`,
			);

			const bootstrapped: Bootstrapped = Bootstrap.fromNodeModules(
				new InFolder(this.paths.nodeModules),
				this.DefaultOptions,
				wantedLibs,
			);

			this.executionLogger.logProgress(
				`Bootstrap completed, found ${bootstrapped.libraryResolver.stdLib.libName.value} as stdlib`,
			);

			// Convert the Scala Either pattern matching to fp-ts Either with fold
			const sources: LibTsSource[] = pipe(
				bootstrapped.initialLibs,
				E.fold(
					(unresolved: Unresolved) => {
						this.executionLogger.logError(
							`Failed to resolve initial libraries: ${unresolved.msg}`,
						);
						throw new Error(unresolved.msg);
					},
					(initial: LibTsSource[]) => initial,
				),
			);

			this.executionLogger.logProgress(
				`Initial sources from bootstrap: ${sources.map((s) => s.libName.value).join(", ")}`,
			);
			this.executionLogger.logProgress(
				`Converting ${sources.map((s) => s.libName.value).join(", ")} to scalajs...`,
			);

			// Execute the three-phase pipeline
			await this.executeThreePhasePipeline(sources, bootstrapped);

			// Finalize execution log on success
			await this.executionLogger.finalizeExecutionLog(true);
		} catch (error) {
			this.failSpinner("Conversion failed");
			await this.executionLogger.finalizeExecutionLog(false);
			throw error;
		}
	}

	/**
	 * Execute the three-phase pipeline: Phase1 → Phase2 → PhaseFlavour
	 */
	private async executeThreePhasePipeline(
		sources: LibTsSource[],
		bootstrapped: Bootstrapped,
	): Promise<void> {
		this.executionLogger.logStep("Setting up three-phase pipeline");

		// Create logger function
		const getLogger = (id: LibTsSource): Logger<void> => {
			return {
				info: (msg: string) => this.info(`[${id.libName.value}] ${msg}`),
				warn: (msg: string) => this.warn(`[${id.libName.value}] ${msg}`),
				error: (msg: string) => this.error(`[${id.libName.value}] ${msg}`),
				debug: (msg: string) =>
					this.info(`[${id.libName.value}] DEBUG: ${msg}`),
				withContext: (_key: string, _value: string) => getLogger(id),
			} as Logger<void>;
		};

		// Create phase listener - use logging listener for better visibility
		const listener = new CollectingPhaseListener<LibTsSource>();

		// Add a logging phase listener for real-time feedback
		const loggingListener = {
			on: (phaseName: string, id: LibTsSource, event: any) => {
				switch (event._tag) {
					case "Started":
						console.log(`🚀 [${phaseName}] Started processing: ${id.libName.value}`);
						break;
					case "Success":
						console.log(`✅ [${phaseName}] Successfully processed: ${id.libName.value}`);
						break;
					case "Failure":
						console.error(`❌ [${phaseName}] Failed to process: ${id.libName.value}`);
						break;
					case "Blocked":
						console.log(`⏸️  [${phaseName}] Blocked on dependencies: ${id.libName.value}`);
						break;
					case "Ignored":
						console.log(`⏭️  [${phaseName}] Ignored: ${id.libName.value}`);
						break;
				}
				// Also forward to collecting listener
				listener.on(phaseName, id, event);
			}
		};

		// Create formatters and orderings
		const formatter = Formatters.create<LibTsSource>(
			(source) => source.libName.value,
		);
		const ordering = Orderings.create<LibTsSource>((a, b) =>
			a.libName.value.localeCompare(b.libName.value),
		);

		// Configure Phase1: TypeScript parsing
		this.executionLogger.logStep("Configuring Phase1ReadTypescript");
		const phase1Config = {
			resolve: bootstrapped.libraryResolver,
			calculateLibraryVersion: new MockCalculateLibraryVersion(),
			ignored: this.DefaultOptions.ignoredLibs,
			ignoredModulePrefixes: this.DefaultOptions.ignoredModulePrefixes,
			pedantic: false,
			parser: () => E.right(TsParsedFile.createMock()), // Mock parser
			expandTypeMappings: this.DefaultOptions.expandTypeMappings,
		};
		const phase1 = Phase1ReadTypescript.create(phase1Config);

		// Configure Phase2: TypeScript to Scala.js conversion
		this.executionLogger.logStep("Configuring Phase2ToScalaJs");
		console.log("🔧 [TracingCommand] Setting up Phase2ToScalaJs configuration...");
		const phase2Config = {
			pedantic: false,
			useDeprecatedModuleNames: this.DefaultOptions.useDeprecatedModuleNames,
			scalaVersion: this.DefaultOptions.versions.scala,
			enableScalaJsDefined: this.DefaultOptions.enableScalaJsDefined,
			outputPkg: this.DefaultOptions.outputPackage,
			flavour: NormalFlavourImpl.createMock(),
		};
		console.log("🏗️  [TracingCommand] Creating Phase2ToScalaJs instance...");
		const phase2 = Phase2ToScalaJs.create(phase2Config);
		console.log("✅ [TracingCommand] Phase2ToScalaJs configured successfully");

		// Configure Phase3: Flavour transformations
		this.executionLogger.logStep("Configuring PhaseFlavour");
		const phase3Config = {
			flavour: NormalFlavourImpl.createMock(),
			maybePrivateWithin: this.DefaultOptions.privateWithin
				? { _tag: "Some" as const, value: this.DefaultOptions.privateWithin }
				: { _tag: "None" as const },
		};
		const phase3 = PhaseFlavour.create(phase3Config);

		// Create the three-phase pipeline
		this.executionLogger.logStep("Creating three-phase pipeline");
		const pipeline = RecPhase.apply<LibTsSource>()
			.next(phase1.apply.bind(phase1), "Phase1ReadTypescript")
			.next(phase2.apply.bind(phase2), "Phase2ToScalaJs")
			.next(phase3.apply.bind(phase3), "PhaseFlavour");

		// Create phase runner - use the logging listener for better visibility
		const runner = PhaseRunner.apply(
			pipeline,
			getLogger,
			loggingListener,
			formatter,
			ordering,
		);

		// Execute pipeline for each source
		this.executionLogger.logStep("Executing three-phase pipeline");
		for (const source of sources) {
			this.executionLogger.logProgress(
				`Processing ${source.libName.value} through three-phase pipeline`,
			);

			const result = runner(source);

			if (result._tag === "Ok") {
				this.executionLogger.logProgress(
					`Successfully processed ${source.libName.value}`,
				);
			} else if (result._tag === "Failure") {
				this.executionLogger.logError(
					`Failed to process ${source.libName.value}: ${Array.from(result.errors.values()).join(", ")}`,
				);
			} else {
				this.executionLogger.logProgress(`Ignored ${source.libName.value}`);
			}
		}

		this.executionLogger.logStep("Three-phase pipeline execution completed");
	}

	private async validateEnvironment(): Promise<void> {
		if (!(await fs.pathExists(this.paths.packageJson))) {
			throw new Error(`${this.inDirectory} does not contain package.json`);
		}

		if (!(await fs.pathExists(this.paths.nodeModules))) {
			throw new Error(`${this.inDirectory} does not contain node_modules`);
		}
	}
}
